R/SCE-reducedDims.R

```diff
@@ -53,25 +53,118 @@ setReplaceMethod("reducedDimNames", c("SingleCellExperiment", "character"), func
 })
 
 #' @export
-setMethod("reducedDim", "SingleCellExperiment", function(x, type=1, withDimnames=TRUE) {
+setMethod("reducedDim", c("SingleCellExperiment", "missing"), function(x, type, withDimnames=TRUE) {
     x <- updateObject(x)
+    internals <- int_colData(x)[[.red_key]]
+
+    if (identical(ncol(internals), 0L)) {
+        .Deprecated(msg="NULL is deprecated.")
+        return(NULL)
+        # To deprecate NULL and throw an error instead, remove the two lines above.
+        stop(
+            "'reducedDim(<", class(x), ">, ...) ",
+            "length(reducedDims(<", class(x), ">)) is 0'")
+    }
+
+    reducedDim(x, 1L, withDimnames)
+})
+
+#' @export
+setMethod("reducedDim", c("SingleCellExperiment", "numeric"), function(x, type=1, withDimnames=TRUE) {
+    x <- updateObject(x)
+    internals <- int_colData(x)[[.red_key]]
+
+    out <- tryCatch({
+        internals[, type]
+    }, error=function(err) {
+        .Deprecated(msg="NULL is deprecated.")
+        return(NULL)
+        # To deprecate NULL and throw an error instead, remove the two lines above.
+        stop("'reducedDim(<", class(x), ">, type=\"numeric\", ...)' ",
+             "invalid subscript 'type'\n", conditionMessage(err))
+    })
+
+    if (withDimnames) {
+        rownames(out) <- colnames(x)
+    }
+
+    out
+})
+
+#' @export
+setMethod("reducedDim", c("SingleCellExperiment", "character"), function(x, type, withDimnames=TRUE) {
+    msg <- paste0("'reducedDim(<", class(x), ">, type=\"character\", ...)' ",
+                  "invalid subscript 'type'")
 
+    x <- updateObject(x)
     internals <- int_colData(x)[[.red_key]]
-    out <- internals[,type]
-    if (!is.null(out) && withDimnames) {
+
+    out <- tryCatch({
+        internals[, type]
+    }, error=function(err) {
+        .Deprecated(msg="NULL is deprecated.")
+        return(NULL)
+        # To deprecate NULL and throw an error instead, remove the two lines above.
+        stop(msg, "\n'", type, "' not in names(reducedDims(<", class(x), ">))")
+    })
+
+    if (withDimnames) {
         rownames(out) <- colnames(x)
     }
+
     out
 })
 
 #' @export
-setReplaceMethod("reducedDim", "SingleCellExperiment", function(x, type=1, ..., value) {
+setReplaceMethod("reducedDim", c("SingleCellExperiment", "missing"), function(x, type, ..., value) {
+    type <- 1L
+
+    if (0L == length(reducedDimNames(x))){
+        ## Implementation 1
+        # https://github.com/drisso/SingleCellExperiment/pull/35#issuecomment-522258649
+        # type <- "unnamed"
+        ## Implementation 2
+        # `reducedDims<-`` above creates character(n) names
+        type <- character(1)
```

Comment:
See Bioconductor/S4Vectors#47 for why we shouldn't be consistent with `assays` on this one.

Reply:
`type <- "unnamed"`